Modules/sha2module.c

```diff
@@ -0,0 +1,789 @@
+/*[clinic input]
+_sha2.sha384
+
+    string: object(c_default="NULL") = b''
+    *
+    usedforsecurity: bool = True
+
+Return a new SHA-384 hash object; optionally initialized with a string.
+[clinic start generated code]*/
+
+static PyObject *
+_sha2_sha384_impl(PyObject *module, PyObject *string, int usedforsecurity)
+/*[clinic end generated code: output=b29a0d81d51d1368 input=4e9199d8de0d2f9b]*/
+{
+    SHA512object *new;
+    Py_buffer buf;
+
+    sha2_state *state = sha2_get_state(module);
+
+    if (string)
+        GET_BUFFER_VIEW_OR_ERROUT(string, &buf);
+
+    if ((new = newSHA384object(state)) == NULL) {
+        if (string)
+            PyBuffer_Release(&buf);
+        return NULL;
+    }
+
+    new->state = Hacl_Streaming_SHA2_create_in_384();
+    new->digestsize = 48;
+
+    if (PyErr_Occurred()) {
+        Py_DECREF(new);
+        if (string)
+            PyBuffer_Release(&buf);
+        return NULL;
+    }
+    if (string) {
+        update_512(new->state, buf.buf, buf.len);
+        PyBuffer_Release(&buf);
+    }
+
+    return (PyObject *)new;
+}
+
+/* List of functions exported by this module */
+
+static struct PyMethodDef SHA2_functions[] = {
+    _SHA2_SHA256_METHODDEF
+    _SHA2_SHA224_METHODDEF
+    _SHA2_SHA512_METHODDEF
+    _SHA2_SHA384_METHODDEF
+    {NULL,      NULL}            /* Sentinel */
+};
+
+static int
+_sha2_traverse(PyObject *module, visitproc visit, void *arg)
+{
+    sha2_state *state = sha2_get_state(module);
+    Py_VISIT(state->sha224_type);
+    Py_VISIT(state->sha256_type);
+    Py_VISIT(state->sha384_type);
+    Py_VISIT(state->sha512_type);
+    return 0;
+}
+
+static int
+_sha2_clear(PyObject *module)
+{
+    sha2_state *state = sha2_get_state(module);
+    Py_CLEAR(state->sha224_type);
+    Py_CLEAR(state->sha256_type);
+    Py_CLEAR(state->sha384_type);
+    Py_CLEAR(state->sha512_type);
+    return 0;
+}
+
+static void
+_sha2_free(void *module)
+{
+    _sha2_clear((PyObject *)module);
+}
+
+/* Initialize this module. */
+static int sha2_exec(PyObject *module)
+{
+    sha2_state *state = sha2_get_state(module);
+
+    state->sha224_type = (PyTypeObject *)PyType_FromModuleAndSpec(
+        module, &sha224_type_spec, NULL);
+    if (state->sha224_type == NULL) {
+        return -1;
+    }
+    state->sha256_type = (PyTypeObject *)PyType_FromModuleAndSpec(
+        module, &sha256_type_spec, NULL);
+    if (state->sha256_type == NULL) {
+        return -1;
+    }
+    state->sha384_type = (PyTypeObject *)PyType_FromModuleAndSpec(
+        module, &sha384_type_spec, NULL);
+    if (state->sha384_type == NULL) {
+        return -1;
+    }
+    state->sha512_type = (PyTypeObject *)PyType_FromModuleAndSpec(
+        module, &sha512_type_spec, NULL);
+    if (state->sha512_type == NULL) {
+        return -1;
+    }
+
+    Py_INCREF((PyObject *)state->sha224_type);
+    if (PyModule_AddObject(module, "SHA224Type", (PyObject *)state->sha224_type) < 0) {
+        Py_DECREF((PyObject *)state->sha224_type);
```

Comment:
If you use `PyModule_AddType`, you can get rid of the unneeded refcount dance here. Ditto for the other types below.

Reply:
this required changing the type tp_name values to match, but that's probably a good idea anyways.